src/main/java/com/redhat/devtools/lsp4ij/features/LSPDeclarativeInlayActionHandler.java

```diff
@@ -0,0 +1,31 @@
+package com.redhat.devtools.lsp4ij.features;
+
+import com.google.common.collect.MapMaker;
+import com.intellij.codeInsight.hints.declarative.InlayActionHandler;
+import com.intellij.codeInsight.hints.declarative.InlayActionPayload;
+import com.intellij.codeInsight.hints.declarative.StringInlayActionPayload;
+import com.intellij.openapi.editor.Editor;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class LSPDeclarativeInlayActionHandler implements InlayActionHandler {
+    //There is no way to send complex data through inlay action payloads, so we cheat a little.
+    public static Map<InlayActionPayload, Consumer<Editor>> callbacks = new MapMaker().weakKeys().makeMap();
+
+    public static InlayActionPayload createAndRegisterPayload(String text, Consumer<Editor> callback) {
+        var payload = new StringInlayActionPayload(text);
```

Comment:
What about extends StringInlayActionPayLoad to stores the callback and avoid using weak map?

Reply:
StringInlayActionPayload is a final class (non-open kotlin class), InlayActionPayload is a sealed interface, and all other classes extending InlayActionPayload are also final classes

Reply:
We *could* use a dynamic proxy style way to create a custom subclass of InlayActionPayload during runtime without tripping the marketplace verifier

Reply:
I don't know how to that and if it is clean?

Main problem is callback will grow up a lot and never clean.

What about using a simple Map stored in the Editor#putUserData?

Reply:
@SCWells72 have you some better idea?

Reply:
Fixed it in the latest push, I'm hijacking the `PsiPointerInlayActionPayload` with a custom subclass of `SmartPsiElementPointer` and using it as the callback carrier